Shortcut guide

tanchekwei opened the pull request:
Add a shortcut guide and some refactoring.

Add shortcut guide and find icon on the top bar.

Add Ctrl + Tab to open bots menu

Comment:

I tried to move the shortcut label using absolute position, but their position changes with different window sizes.

I have been trying for hours to find the reason but with no luck. Fortunately, I found using canvas to draw SVG can keeps the shortcut label position consistent regardless of the window size.

Limitation:
Due to draw with the canvas, the image is not sharp, I not sure how to increase the resolution yet.

The favorite bot shortcut label is not aligned when the window is small.


![image](https://github.com/sunner/ChatALL/assets/26683979/c35ae7f8-b735-49ef-84a8-f9ec8797213d)

sunner commented:
Wow, amazing work👍

I believe that what I said yesterday misleads you. I said on macOS `Cmd + F` is usually for finding, that means it is better to change the shortcut from Ctrl to Cmd in the code. But forget it, I tried, the Cmd key (`meta` in vue-shortkey) does not work.

The `ctrl` does work on macOS. So, you need not to draw different key name on the guide. It must waste some of your time. Sorry.

Another suggest. Could you move the question icon to the far right? Only the frequency used ones should be on the left.

tanchekwei commented:
> Wow, amazing work👍
> 
> I believe that what I said yesterday misleads you. I said on macOS `Cmd + F` is usually for finding, that means it is better to change the shortcut from Ctrl to Cmd in the code. But forget it, I tried, the Cmd key (`meta` in vue-shortkey) does not work.
> 
> The `ctrl` does work on macOS. So, you need not to draw different key name on the guide. It must waste some of your time. Sorry.
> 
> Another suggest. Could you move the question icon to the far right? Only the frequency used ones should be on the left.

Thanks for your review. @sunner 

Moved question icon to far right. [41a18c8](https://github.com/sunner/ChatALL/pull/272/commits/41a18c859ca7f58a11441b2214a1449fc64d436f)

So I'll just show `Ctrl` in the guide for all OS right? If that's right, I'll remove those code checking user OS for showing either Cmd or Ctrl.

I changed PR to draft first because I found that `v-overlay` is the issue with positioning elements. I am going to remove those SVG and canvas drawings because they are ugly 😂

sunner commented:
> So I'll just show `Ctrl` in the guide for all OS right? If that's right, I'll remove those code checking user OS for showing either Cmd or Ctrl.

Sure, show `Ctrl` on macOS is also right.

tanchekwei commented:
Done @sunner kindly review. Thanks

tanchekwei commented:
Done @sunner kindly review. Thanks

sunner commented:
Thank you @tanchekwei . This is a really great PR! I need some time to cherry-pick independent commits and combine related commits to keep the history of main branch clear.

tanchekwei commented:
> Thank you @tanchekwei . This is a really great PR! I need some time to cherry-pick independent commits and combine related commits to keep the history of main branch clear.

Sure, let me know if you want me to combine those commits.

And should I resolve the conflict.

sunner commented:
I've cherry-picked and released these commits in v1.29.37 :

```git
commit 96d9766a826b2eaa0cfef810962489f47787cb83 (HEAD -> main, tag: v1.29.37, origin/main, origin/HEAD)
Author: tanchekwei <weiwei1@live.com.my>
Date:   Sat Jun 24 15:46:32 2023 +0800

    Add Ctrl + Tab to toggle bots menu

commit 5d1bc3d854213b84e0c33dcb80bc09dc6dd43c98
Author: tanchekwei <weiwei1@live.com.my>
Date:   Sat Jun 24 15:42:44 2023 +0800

    Use v-for to create column icons

commit 73dddbdbe3a3f1c333a2b3e942eca6b9de0a5f12
Author: tanchekwei <weiwei1@live.com.my>
Date:   Sat Jun 24 15:41:44 2023 +0800

    Add try catch as hit unknown error on JSON parse

commit 94f88cdb555d946b919991f86930fcb5fab2aa3f
Author: tanchekwei <weiwei1@live.com.my>
Date:   Sat Jun 24 15:44:56 2023 +0800

    Add find icon
```

The `Use v-for to create column icons` commit is combined with [Update getColumnSvg naming](https://github.com/sunner/ChatALL/pull/272/commits/a918810b9d2aac8a94560925f4696b5468f02345)

The remain commits in this PR are all related with Shortcut Guide and constant things if I don't miss anything. 

Unfortunately, they are conflicted with the main branch. I tried to resolve but can't understand all things well. Could you rebase the repo with the origin/main first, and then `git rebase -i xxxx` to combine them as one commit? (It is better to make the guide and constant as two commits, if it is convenient.)

Thank you.

tanchekwei commented:
@sunner 

May I know how to rebase the repo with the origin/main? Sorry I don't normally rebase with a branch.

I am more familiar with squash commit using `git rebase -i HEAD~X`

tanchekwei commented:
Ok done

tanchekwei commented:
Done @sunner, tested good as well

sunner commented:
Wow, so good!

Next time, if you like, you can submit different PRs based on different branches for different proposals. This is the typical workflow which can make our life easier. :-)

tanchekwei commented:
> Wow, so good!
> 
> Next time, if you like, you can submit different PRs based on different branches for different proposals. This is the typical workflow which can make our life easier. :-)

Ok sure.
